light-client/relayer/src/relayer.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2018-2023 The MobileCoin Foundation
+
+use crate::{Config, Sender};
+use mc_account_keys::burn_address_view_private;
+use mc_blockchain_types::{BlockData, BlockIndex, BlockMetadata};
+use mc_common::logger::{log, Logger};
+use mc_ledger_db::{Error as LedgerError, Ledger, LedgerDB};
+use mc_transaction_core::tx::TxOut;
+use mc_util_telemetry::{block_span_builder, telemetry_static_key, tracer, Key, Span, Tracer};
+use mc_watcher::{error::WatcherDBError, watcher_db::WatcherDB};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use std::time::Duration;
+
+/// Telemetry: block index currently being worked on.
+const TELEMETRY_BLOCK_INDEX_KEY: Key = telemetry_static_key!("block-index");
+
+/// The relayer object is able to scan the blockchain for interesting burn txos,
+/// and forward them to a "sender", together with proofs of the block's
+/// validity, when it finds any.
+pub struct Relayer<S: Sender> {
+    config: Config,
+    next_block_index: BlockIndex,
+    ledger_db: LedgerDB,
+    watcher_db: WatcherDB,
+    sender: S,
+    logger: Logger,
+}
+
+impl<S: Sender> Relayer<S> {
+    /// Poll for new data every 10 ms
+    const POLLING_FREQUENCY: Duration = Duration::from_millis(10);
+    /// How frequently to retry if an error occurs.
+    const ERROR_RETRY_FREQUENCY: Duration = Duration::from_millis(1000);
+
+    /// Create a new relayer
+    pub fn new(
+        config: Config,
+        ledger_db: LedgerDB,
+        watcher_db: WatcherDB,
+        sender: S,
+        logger: Logger,
+    ) -> Self {
+        let next_block_index = config.start_block_index;
+        Self {
+            config,
+            next_block_index,
+            ledger_db,
+            watcher_db,
+            sender,
+            logger,
+        }
+    }
+
+    /// Entrypoint for relayer loop.
+    pub fn entry_point(&mut self) -> ! {
+        // Poll ledger for data to relay
+        loop {
+            match self.ledger_db.get_block_data(self.next_block_index) {
+                Err(LedgerError::NotFound) => std::thread::sleep(Self::POLLING_FREQUENCY),
+                Err(e) => {
+                    log::error!(
+                        self.logger,
+                        "Unexpected error when checking for block data {}: {:?}",
+                        self.next_block_index,
+                        e
+                    );
+                    std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                }
+                Ok(block_data) => {
+                    if let Err(err) = self.process_block(&block_data) {
+                        log::error!(
+                            self.logger,
+                            "When processing block {}: {:?}",
+                            self.next_block_index,
+                            err
+                        );
+                        std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
```

Comment:
:question: If I'm understanding the behavior correctly this will only error when `check_for_signatures()` errors, which only happens if `watcher_db.get_block_data_map()` errors.  Doing a quick surface level check of the errors in `get_block_data_map()`, they don't appear to be recoverable, which means this is likely to be an infinite loop.

Is it intentional that this will continue to run and log unrecoverable errors?

Similar thing for the `Err(e)` handling up above fro `ledger_db.get_block_data()`

Reply:
I think this loop should be okay here.

Blocks that are actually externalized by the network have to be approved by enough nodes, and so we should eventually get enough signatures for those blocks if we loop.

If the network forks, or mobilecoind puts a block in the ledger that wasn't actually externalized by a quorum, that would be bad, but I think it's okay to require intervention in that case. I don't think that has ever happened in practice or testing.